Add a hook to track packets not processed

timwoj opened the pull request:
This PR adds a new hook called `packet_not_processed` that can be invoked whenever a packet is received that wasn't processed by an analyzer. The heuristic for "not processed" currently just means "nothing was written to a log file for this packet". Since the majority of our analyzers end up writing an entry to `conn.log`, this is the starting point for what is considered processed. We can expand on it later.

It also adds a new command-line option to dump unprocessed packets a file, called `-c` or `--capture-unprocessed`. This should help aid in debugging new analyzers or for capturing traffic that's evading zeek's analyzers entirely.

Closes #1620

0xxon commented on src/packet_analysis/Manager.h:
```diff
@@ -40,7 +46,7 @@ class Manager : public plugin::ComponentManager<Tag, Component>
 	 * Second-stage initialization of the manager. This is called late
 	 * during Zeek's initialization after any scripts are processed.
```
This should probably get an explanation of the argument in the documentation.

timwoj replied:
Fixed, and also fixed it to not take a `std::optional` but just a string directly.

0xxon commented on src/packet_analysis/Manager.h:
```diff
@@ -126,6 +132,8 @@ class Manager : public plugin::ComponentManager<Tag, Component>
 		return pkt_filter;
 		}
 
+	uint64_t GetUnprocessedCount() const { return total_not_processed; }
```
This is missing a documentation header :)

timwoj replied:
Added

0xxon commented on src/packet_analysis/Manager.h:
```diff
@@ -163,6 +171,10 @@ class Manager : public plugin::ComponentManager<Tag, Component>
+
+	FuncPtr packet_not_processed_hook = nullptr;
+	uint64_t total_not_processed = 0;
+	iosource::PktDumper* unprocessed_dumper = nullptr;
```
Mostly curious - is this deleted during shutdown? (I know, there is a ton of stuff that is not - but I wonder if, for new things, we should).

This might honestly not be worth addressing though.

timwoj replied:
The IOSource manager keeps track of all of the open dumpers and destroys them when it shuts down.

0xxon reviewed (approved):
Besides the small nits that I noted - this looks good to me (though, perhaps @rsmmr also wants to tale a look).

rsmmr commented:
Just took a quick look so far. The logic looks good, but I think we have a misunderstanding (or better: ambiguity) what "hook" means. The PR adds a script-level hook so that Zeek scripts get to see these packets. The feature request was a *plugin* hook, so that external C++ plugins get access to them. That requires adding a new hook to `plugin/Plugin.{h,cc}` and `plugin/Manager.{h,cc}`.

We could keep the script-level hook, too, but actually I'm not sure it'd be viable to use it in most situations: depending on environment, there might be a quite a number of unprocessed packets and sending them all through script-land can get expensive. Plus, this type of packets is something that could hit suddenly (say, a new device on the network speaks an unknown encapsulation) and suddenly the script-side hook would be called a lot.

timwoj commented:
> The feature request was a plugin hook, so that external C++ plugins get access to them. That requires adding a new hook to plugin/Plugin.{h,cc} and plugin/Manager.{h,cc}.

Added in https://github.com/zeek/zeek/pull/1843/commits/e6df2cdbfff6427cdbd618f1d45ffaf032e6d9d1

rsmmr commented on src/packet_analysis/Manager.cc:
```diff
@@ -106,6 +108,12 @@ void Manager::ProcessPacket(Packet* packet)
+	if ( ! packet->processed && packet_not_processed_hook )
+		{
+		packet_not_processed_hook->Invoke(Packet::ToVal(packet));
+		total_not_processed++;
```
Should this counter increase really be tied to the having the hook implemented?

rsmmr commented on src/plugin/Plugin.h:
```diff
@@ -357,6 +360,15 @@ class HookArgument
 		}
 
+	/**
+	 * Constructor with a zeek::Args argument.
```
`s/zeek::Args/packet/`

rsmmr commented on src/plugin/Plugin.h:
```diff
@@ -506,6 +518,16 @@ class HookArgument
 		}
 
+	/**
+	 * Returns the value for a vod pointer argument. The argument's type
```
`s/vod/packet/`?

rsmmr commented on scripts/base/init-bare.zeek:
```diff
@@ -5496,4 +5496,8 @@ event net_done(t: time)
 
+## A hook taking a packet timestamp. Called when a packet reaches the end of
+## processing without having been marked processed.
+global packet_not_processed: hook(pkt: pcap_packet);
```
Now that I'm thinking about this, in script-land this should probably be an event, not a hook; similar to `new_packet`. At some level, I am still wondering if we should offer this at all. I'm fine keeping it, but woudkl suggest to add a note to the comment that handling this could be expensive if there's lots of unprocessed. traffic

timwoj replied:
I think removing it is fine. We don't explicitly do this for the other types of plugin hooks, and if someone really wants the information from it they can implement the hook.

rsmmr reviewed (commented):
This is nice!

The one thing missing is a test: would be good to add one similar to the other plugin hook tests.

I'm only seeing one `pkt->processed = true;` for IP. Should ARP get one? (Or did you leave the question where/whether to put confirmations elsewhere as something still to be examined further?)

timwoj commented:
> The one thing missing is a test: would be good to add one similar to the other plugin hook tests.

I added the extra bits to the `plugins.hooks` test, but that test just uses the `http/get.trace` file, and all of the packets in that trace are processed. I can add a second invocation of zeek to that test that reads one of the ARP files, assuming we don't fix that one.

> I'm only seeing one pkt->processed = true; for IP. Should ARP get one? (Or did you leave the question where/whether to put confirmations elsewhere as something still to be examined further?)

The latter. We definitely do process ARP, so there's really not a good reason to leave it out, but there's definitely questions about what other places to mark things as processed.

timwoj commented:
This went through a few changes, if you want to re-review it:

- Changed the script-level hook to an event
- Squashed the hook and event commits down into one
- Added a btest that exercises the hook, the event, and the packet dumping all in one
- Added a NEWS entry
- Rebased it on master again because that other hook PR merged and was causing conflicts

rsmmr commented on NEWS:
```diff
@@ -40,6 +40,15 @@ New Functionality
 
+- Added a new plugin hook for capturing packets that made it through analysis
+  without being processed called ``Plugin::HookUnprocessedPacket``. Currently
+  only ARP packets or packets with a valid IP-based transport header are marked
```
Would remove the `only` because it makes it sounds like there's something substantial missing.
